rust/lance/src/io/reader.rs

```diff
@@ -289,7 +283,62 @@ impl FileReader {
             fragment_id,
             with_row_id: false,
             deletion_vector,
+            stats_page_table,
+        })
+    }
+
+    async fn read_metadata(
+        object_reader: &dyn ObjectReader,
+        session: Option<&Session>,
+    ) -> Result<Arc<Metadata>> {
+        Self::load_from_cache(session, object_reader.path(), |_| async {
+            let file_size = object_reader.size().await?;
+            let begin = if file_size < object_reader.block_size() {
+                0
+            } else {
+                file_size - object_reader.block_size()
+            };
+            let tail_bytes = object_reader.get_range(begin..file_size).await?;
+            let metadata_pos = read_metadata_offset(&tail_bytes)?;
+
+            let metadata: Metadata = if metadata_pos < file_size - tail_bytes.len() {
+                // We have not read the metadata bytes yet.
+                read_struct(object_reader, metadata_pos).await?
+            } else {
+                let offset = tail_bytes.len() - (file_size - metadata_pos);
+                read_struct_from_buf(&tail_bytes.slice(offset..))?
+            };
+            Ok(metadata)
         })
+        .await
+    }
+
+    /// Get the statistics page table. This will read the metadata if it is not cached.
+    ///
+    /// The page table is cached.
+    async fn read_stats_page_table(
+        object_reader: &dyn ObjectReader,
+        session: Option<&Session>,
+    ) -> Result<Arc<Option<PageTable>>> {
+        // To prevent collisions, we cache this at a child path
+        Self::load_from_cache(session, &object_reader.path().child("stats"), |_| async {
```

Comment:
Do we do this kind of independent caching elsewhere?  Is the id basically to allow us to cache the stats independent of the data?

Reply:
We don't. The key is cached with `(path, TypeId)`, and there is already a `PageTable` for the `path` (the one pointing to the actual data), so I felt that it would be good to use a different `path` to store the stats page table.